### DISCUSSION TITLE
Image Sequence to APNG

### DISCUSSION THREAD

**Discussion by Jay123210599** [ImageMagick/ImageMagick#7767] [Help]
### ImageMagick version

7.1.1-40

### Operating system, version and so on

Windows 11

### Description

How do I create APNGs out of image sequences?

**Comment by fmw42** [comment]
`magick image1.suffix image2.suffix ... imageN.suffix APNG:result.png`

**Reply by Jay123210599** [reply]
Will the result be lossless?

**Reply by fmw42** [reply]
You must also have ffmpeg installed.

**Reply by Jay123210599** [reply]
I did through chocolatey. Do I need to make APNGs at all?

**Reply by fmw42** [reply]
Sorry, you have not explained what you are doing?  You only asked how to make them?

**Reply by Jay123210599** [reply]
I'm asking how to make lossless APNGs out of image sequences.

**Reply by fmw42** [reply]
I am not sure how to make lossless ones.  PNG is lossless, so I assume the APNGs will be also.  A developer or someone who can look at the software may be able to comment.  Since they are created with ffmpeg, you might review ffmpeg documentation.

**Comment by snibgo** [comment]
As fmw42 says, writing APNG is delegated to the external delegate ffmpeg, using the `encode="video:encode"` entry in `delegates.xml`, rather than the internal delegate libpng.

The output APNG is always 8 bits, even if we insert `-sample_fmt u16` or `-sample_fmt s16` into the ffmpeg command. I don't know if this is a limitation of APNG, or a limitation within IM that can be fixed.

So input images that are more than 8 bits per channel will lose precision.

If we don't have `-quality 100`, then output APNG images will be different to the input images (eg RMSE 2%). I suppose this is a lossy compression.

If we do have `-quality 100`, then output APNG images will be identical to the 8 bit/channel inputs. Like this:
```
magick {input_files} -quality 100 apng:out.png
```

**Reply by Jay123210599** [reply]
how do you know all of that?

**Comment by snibgo** [comment]
`-verbose` shows the ffmpeg command. `exiftool x.png` shows the bits per channel. Editing `delegates.xml` to include `-sample_fmt u16` etc doesn't improve the bits per channel. Experimenting with `-quality N` and using `magick compare` shows whether images are identical or different.

**Reply by Jay123210599** [reply]
How do I use verbose and magick compare?

**Comment by snibgo** [comment]
An example of `-verbose`:
```
magick -verbose toes.png toes.png apng:x.apng

toes.png PNG 267x233 267x233+0+0 16-bit TrueColor sRGB 320268B 0.003u 0:00.003
toes.png PNG 267x233 267x233+0+0 16-bit TrueColor sRGB 320268B 0.003u 0:00.003
toes.png=>C:/Users/Alan/AppData/Local/Temp/magick-BMXfzOCb7TDvC56rpTwogic3Ccdiu-wo0.webp PNG 267x233 267x233+0+0 16-bit TrueColor sRGB 5200B 0.008u 0:00.007
toes.png=>C:/Users/Alan/AppData/Local/Temp/magick-BMXfzOCb7TDvC56rpTwogic3Ccdiu-wo1.webp[1] PNG 267x233 267x233+0+0 16-bit TrueColor sRGB 5200B 0.007u 0:00.006
"ffmpeg.exe" -nostdin -loglevel error -sample_fmt u16 -i "C:/Users/Alan/AppData/Local/Temp/magick-BMXfzOCb7TDvC56rpTwogic3Ccdiu-wo%d.webp" -plays 0 "C:/Users/Alan/AppData/Local/Temp/magick-xmNVdv_xO0jamstbQ_NWTlUe3U3J9D7y.APNG"
toes.png=>x.apng[0] PNG 267x233 267x233+0+0 16-bit TrueColor sRGB 320268B 0.098u 0:00.097
```
The text output shows:
- Reading two inputs.
- Writing two webp files (these are intermediate files that will be sent to ffmpeg).
- The ffmpeg command. Note that this uses the outputs from the previous step as input (`-i`).
- Writing the final output to x.apng.

An example of `magick compare`:
```
magick compare -metric RMSE toes.png apng:x.apng[0] NULL:
1251.14 (0.0190912)
```
This compares `toes.png` with the first image (counting from zero) in `x.apng`. The RMSE is 0.0190912 on a scale of 0 to 1, so about 1.9%. `toes.png` is 16 bits per channel, and the first image in `x.apng` is not identical to `toes.png`.

**Reply by Jay123210599** [reply]
How do I use compare to show whether two APNGs are identical or different? How would I know if they are identical? They are both 8 bit per channel, but the file sizes are different.

**Comment by snibgo** [comment]
Assuming the APNG files have the same number of frames, all the same size (width and height), and you want to compare corresponding frames of the two files: you could extract all the frames into lossless files (eg PNG), and compare each pair.

For example (Windows BAT syntax), we compare files `x1.apng` and `x2.apng`. We set `numFrames` to the number of frames, then split the frames into two lists. In this example, I use leading zeroes in filenames. This isn't necessary for your needs, but is good practice. We loop through each pair of frames, comparing them.
```
setlocal enabledelayedexpansion

for /F "usebackq" %%L in (`magick apng:x1.apng -format "numFrames=%%N\n" info:`) do set %%L

magick apng:x1.apng x1-%%06d.png
magick apng:x2.apng x2-%%06d.png

echo numFrames=%numFrames%

echo off

for /L %%N in (1,1,%numFrames%) do (
  set /A Ndx=%%N-1
  set LZ=00000!Ndx!
  set LZ=!LZ:~-6!
  magick compare ^
    -metric RMSE ^
    x1-!LZ!.png x2-!LZ!.png ^
    NULL:
  echo.
)
```
The text output when the two APNG files have three frames each might be:
```
0 (0)
2983.46 (0.0455246)
4429.01 (0.0675823)
```
The first pair of frames are identical, the next pair have 4.6% RMSE, the final pair have 6.8% RMSE.

EDIT: Added line "setlocal enabledelayedexpansion" at top.

**Reply by Jay123210599** [reply]
So they're not identical in this example?

**Reply by Jay123210599** [reply]
I tried out the script but I got this error. How do I fix it?
![Desktop 2024 11 12 - 08 45 41 01 mp4_20241112_085910 870](https://github.com/user-attachments/assets/294a2dd7-8862-41ce-be1b-ba725ac7bf2c)

**Reply by snibgo** [reply]
My script above may need "setlocal enabledelayedexpansion" (without quotes) in a line inserted at the top.

**Reply by Jay123210599** [reply]
How do I do that?

**Comment by fmw42** [comment]
@snibgo You wrote:

> magick {input_files} -quality 100 apng:out.png

How do you know that -quality 100 is appropriate for APNG.   PNG files do not use that value (100) for best quality. So is APNG different from PNG in compression quality?

**Comment by snibgo** [comment]
Yes, "-quality" for video formats (including APNG) has a different meaning to "-quality" for PNG format.

I have discovered that to get 16 bit/channel output APNG, we can use:
```
-define video:pixel-format=rgb48be
```

**Reply by Jay123210599** [reply]
So what does that mean? `-quality 100` is not needed making for lossless APNGs?

**Comment by snibgo** [comment]
If you want lossless APNGs, use "-quality 100".

**Reply by Jay123210599** [reply]
The script now works, but the results only show up for a short while before the script exists. How do I save the results to a txt file or something else?

**Comment by snibgo** [comment]
> ... the results only show up for a short while before the script exists.

Insert a "pause" command at the end of the script. Better, open a "Command Prompt" (aka "console") window, and type the name of the BAT file there.

> How do I save the results to a txt file or something else?

Redirect the output to a file using `>` or `>>`.

**Reply by Jay123210599** [reply]
> > ... the results only show up for a short while before the script exists.
> 
> Insert a "pause" command at the end of the script. Better, open a "Command Prompt" (aka "console") window, and type the name of the BAT file there.

Can you add that to the script? 

> > How do I save the results to a txt file or something else?
> 
> Redirect the output to a file using `>` or `>>`.

Can you show me an example?

**Comment by snibgo** [comment]
This forum is about ImageMagick. For the basics of Windows shell scripting, I suggest you look elsewhere, such as Microsoft documentation.

**Comment by Jay123210599** [comment]
Anyway, here are the results. x1.apng is 335MB without `-quality 100`, and x2.apng is 570MB with it.

```
913.909 (0.0139454)
18842.1 (0.287512)
16992.1 (0.259282)
19154.2 (0.292274)
17567.6 (0.268064)
18989.4 (0.289759)
17446.8 (0.26622)
18656.3 (0.284676)
14786.8 (0.225632)
20383.4 (0.31103)
14366.4 (0.219216)
20433.7 (0.311798)
15773.8 (0.240693)
20507.9 (0.31293)
17564.7 (0.268019)
20888.5 (0.318739)
17479.6 (0.266721)
22014.7 (0.335923)
18495.8 (0.282227)
21411.1 (0.326712)
17360.5 (0.264905)
19926.1 (0.304053)
16825.4 (0.256739)
20746.4 (0.31657)
19639.2 (0.299676)
20511.4 (0.312984)
18063.1 (0.275625)
20755.5 (0.316709)
17777.4 (0.271266)
20346.2 (0.310463)
17991.6 (0.274534)
19542 (0.298192)
18250.1 (0.278479)
19386.7 (0.295821)
18465.9 (0.281771)
18471.4 (0.281855)
17941.2 (0.273765)
19357.5 (0.295377)
18120.2 (0.276496)
19572.4 (0.298655)
18096.9 (0.276141)
19606.9 (0.299182)
19109 (0.291585)
19751.3 (0.301386)
17656 (0.269414)
20262.6 (0.309188)
16047.8 (0.244874)
4225.94 (0.0644838)
32954.2 (0.502849)
31026.9 (0.47344)
36595.6 (0.558412)
21022.4 (0.320781)
24977.9 (0.381138)
11924.2 (0.181951)
19153.8 (0.292269)
11249.8 (0.171662)
15648 (0.238774)
14601.9 (0.22281)
28303 (0.431877)
28641.9 (0.437047)
34569.2 (0.527492)
32171.6 (0.490907)
35028.8 (0.534505)
35244.2 (0.537792)
34554.4 (0.527266)
37456.5 (0.571549)
36360.3 (0.554823)
35819.3 (0.546568)
36641.2 (0.55911)
34216.4 (0.522108)
1956.11 (0.0298484)
39435.5 (0.601746)
30798.7 (0.469958)
40633.8 (0.620032)
28036.6 (0.427811)
43608.4 (0.665422)
17751.6 (0.270872)
39012.2 (0.595288)
22687.3 (0.346186)
38648.2 (0.589733)
10587.1 (0.161549)
23724.3 (0.36201)
13109.8 (0.200043)
21583.7 (0.329347)
15950.2 (0.243385)
22163.6 (0.338195)
24397.9 (0.372288)
26960.8 (0.411396)
27831.3 (0.424678)
25953.5 (0.396026)
28632.6 (0.436906)
29027.2 (0.442927)
25470.7 (0.388658)
3308.14 (0.0504791)
16261.2 (0.248129)
16300.2 (0.248725)
13426.5 (0.204876)
17611.7 (0.268738)
18059.9 (0.275576)
15095.7 (0.230345)
16556 (0.252629)
16754.6 (0.255658)
14523.4 (0.221612)
16342.1 (0.249365)
16389 (0.250079)
19052.5 (0.290722)
20848 (0.31812)
20584.8 (0.314103)
14999.9 (0.228884)
16870 (0.25742)
17048.1 (0.260137)
15418.5 (0.235271)
17151 (0.261707)
17346.8 (0.264695)
14664.4 (0.223764)
17903.1 (0.273185)
15196.8 (0.231888)
17905.5 (0.273221)
14014.1 (0.213842)
16486 (0.25156)
14045.6 (0.214321)
15442.3 (0.235634)
12629.4 (0.192713)
19721.6 (0.300932)
14435.3 (0.220269)
21160.5 (0.322888)
16254.8 (0.248032)
23062.5 (0.351911)
22737.7 (0.346955)
17113.3 (0.261132)
20314.7 (0.309982)
16735.2 (0.255363)
12859.1 (0.196217)
13612.5 (0.207714)
12741.2 (0.194419)
11701.1 (0.178547)
13775.8 (0.210205)
15511 (0.236683)
15770.6 (0.240644)
505.871 (0.00771909)
4176.14 (0.0637239)
1701.37 (0.0259613)
2522.21 (0.0384865)
784.111 (0.0119648)
788.534 (0.0120323)
772.075 (0.0117811)
1483.65 (0.0226391)
5665.45 (0.0864492)
21564.1 (0.329047)
30009.4 (0.457914)
31807.5 (0.485352)
30558.1 (0.466287)
31342.2 (0.478251)
32489.8 (0.495763)
39061.4 (0.596039)
40017.2 (0.610623)
39744.7 (0.606465)
39259.6 (0.599064)
36270.7 (0.553456)
35096 (0.535531)
39712.2 (0.605969)
40428 (0.616892)
39928.9 (0.609276)
39646.1 (0.604961)
37034.5 (0.56511)
35728.6 (0.545183)
39516.6 (0.602985)
39705.6 (0.605869)
39619 (0.604547)
40223.6 (0.613772)
37625.7 (0.574132)
36147.5 (0.551575)
39323.4 (0.600036)
39271.9 (0.59925)
38930.4 (0.594039)
37934 (0.578835)
36377.7 (0.555088)
```

So, are they identical or different?

**Reply by fmw42** [reply]
Use magick compare to see the difference.

**Reply by Jay123210599** [reply]
This is what I got: `21577 (0.329244)`

**Comment by snibgo** [comment]
> So, are they identical or different?

None of the scores are zero, so every pair of corresponding frames is different.

**Reply by Jay123210599** [reply]
So which one is lossless?

**Comment by snibgo** [comment]
From just that list of numbers, we can't know whether one APNG is lossless, or which one that is. If you have the images that are the source of both APNG files, then you can compare each of those images with the frames of the first APNG, and then with the second APNG.

**Reply by Jay123210599** [reply]
They both came from the same source of images. I used the same image sequence to make both APNG files.

**Comment by Jay123210599** [comment]


**Reply by fmw42** [reply]
And what is the question?
